### PR TITLE
Allow adding and removing extensions from Dev UI

### PR DESCRIPTION
**Pull request by phillip-kruger**
This PR add the "quarkus add" and "quarkus remove" to Dev UI. It allows you to browse (and search) for installable extensions in Dev UI and install them from Dev UI. It also adds a remove button on the details screen on the extension if it can be uninstalled.

extension_management_dev_ui

### CODE REVIEW

**Review by melloware — APPROVED** [review]

**Comment by cescoffier** [comment]
Looks like it breaks a few things :-)

**Comment by phillip-kruger** [comment]
@cescoffier yes I don;t know why. The tests that breaks in CI runs fine on localhost. Maybe @maxandersen can comment ?

**Review by gsmet — CHANGES_REQUESTED** [review]
I added a micro comment but, more importantly, before merging, I think we need to make sure this doesn't slow down the startup too much:

- Creating an `ObjectMapper` can be sometimes slow. Now it might just be the first `ObjectMapper` creation that is slow but we need to check for it.
- The second thing that worries me is that the test failures seem to point out the fact that you're trying to update the registry even when not actually doing anything related to your new feature. Updating the registry is slow so I definitely don't want us to pay the price of updating it every day if you don't need it - and there's a good chance you won't add a dependency or create a new app every day. To test that, you will have to drop the local registry artifact and see how it performs at startup.

Not saying these are actual issues or issues we cannot solve but we need to make sure this feature is not introducing a regression in daily life developer experience.

This review can be dropped once everything is sorted out but I just want to make sure nobody merges before we have checked the situation.

**Comment by phillip-kruger** [comment]
The test broke because of Remove Dev. This is now fixed, let's see what CI does

File: extensions/vertx-http/dev-ui-resources/src/main/resources/dev-ui/qwc/qwc-extension.js [hunk 1]
```diff
@@ -260,9 +267,30 @@ export class QwcExtension extends LitElement {
...
+            return html`<vaadin-button style="width: 100%;" theme="secondary error" @click="${this._uninstall}">
+                        <vaadin-icon icon="font-awesome-solid:trash-can" slot="prefix"></vaadin-icon>
+                        Remove this extension
+                    </vaadin-button>`;
+        }
+    }
+
+    _uninstall(){
+        this._dialogOpened = false;
+        notifier.showInfoMessage(this.name + " removal in progess");
```
Comment (gastaldi) [inline extensions/vertx-http/dev-ui-resources/src/main/resources/dev-ui/qwc/qwc-extension.js:285]:
```suggestion
        notifier.showInfoMessage(this.name + " removal in progress");
```

File: extensions/vertx-http/dev-ui-resources/src/main/resources/dev-ui/qwc/qwc-extensions.js [hunk 2]
```diff
@@ -243,5 +279,55 @@ export class QwcExtensions extends observeState(LitElement) {
...
+                        <vaadin-icon class="addExtensionIcon" icon="font-awesome-solid:plus"></vaadin-icon>
+                    </vaadin-button>`;
+        }
+    }
+    
+    _installRequest(e){
+        this._addDialogOpened = false;
+        let name = e.detail.name;
+        if(e.detail.outcome){
+            notifier.showInfoMessage(name + " instalation in progess");
```
Comment (gastaldi) [inline extensions/vertx-http/dev-ui-resources/src/main/resources/dev-ui/qwc/qwc-extensions.js:322]:
```suggestion
            notifier.showInfoMessage(name + " installation in progress");
```

File: extensions/vertx-http/dev-ui-resources/src/main/resources/dev-ui/qwc/qwc-extensions.js [hunk 3]
```diff
@@ -243,5 +279,55 @@ export class QwcExtensions extends observeState(LitElement) {
...
+                () => html`<qwc-extension-add @inprogress="${this._installRequest}"></qwc-extension-add>`
+              )}
+            ></vaadin-dialog>
+            ${this._renderAddExtensionButton()}
+          `;
+    }
+    
+    _renderAddExtensionButton(){
+        if(connectionState.current.isConnected){
+            return html`<vaadin-button class="addExtensionButton" theme="icon" aria-label="Add extension" title="Add extension" @click="${this._openAddDialog}">
```
Comment (gastaldi) [inline extensions/vertx-http/dev-ui-resources/src/main/resources/dev-ui/qwc/qwc-extensions.js:312]:
```suggestion
            return html`<vaadin-button class="addExtensionButton" theme="icon" aria-label="Add Extension" title="Add Extension" @click="${this._openAddDialog}">
```

File: independent-projects/tools/devtools-common/src/main/java/io/quarkus/devtools/commands/handlers/ListCategoriesCommandHandler.java [hunk 4]
```diff
@@ -20,6 +25,7 @@ public class ListCategoriesCommandHandler implements QuarkusCommandHandler {
     private static final String ID_FORMAT = "%-20s";
     private static final String CONCISE_FORMAT = "%-50s %-50s";
     private static final String FULL_FORMAT = "%-30s %-20s %s";
+    private ObjectMapper objectMapper = new ObjectMapper();
```
Comment (gastaldi) [inline independent-projects/tools/devtools-common/src/main/java/io/quarkus/devtools/commands/handlers/ListCategoriesCommandHandler.java:28]: I agree with @gsmet here, I am bit reluctant to instantiate a new ObjectMapper here. I believe we have already done that somewhere else in the code

**Review by gastaldi — CHANGES_REQUESTED** [review]

File: independent-projects/tools/devtools-common/src/main/java/io/quarkus/devtools/commands/handlers/ListCategoriesCommandHandler.java (hunk 4 above)
Comment (phillip-kruger) [reply independent-projects/tools/devtools-common/src/main/java/io/quarkus/devtools/commands/handlers/ListCategoriesCommandHandler.java:28]: Sure. Where ? Or tell me how I can get an instance?

Comment (phillip-kruger) [reply independent-projects/tools/devtools-common/src/main/java/io/quarkus/devtools/commands/handlers/ListCategoriesCommandHandler.java]: I removed the usage of json and just use Object, so no need anymore to go to json

**Comment by phillip-kruger** [comment]
@gsmet I am not sure how to check this. You mentioned "drop the local registry artifact" - how do I do that ? I cleared my .quarkus and .jbang folders in home, but all seems fine. Any tips ?

**Review by gastaldi — APPROVED** [review]

**Comment by gastaldi** [comment]
> @gsmet I am not sure how to check this. You mentioned "drop the local registry artifact" - how do I do that ? I cleared my .quarkus and .jbang folders in home, but all seems fine. Any tips ?

I believe @gsmet is talking about the `~/.m2/repository/io/quarkus/registry/quarkus-non-platform-extensions/1.0-SNAPSHOT/` artifact that is downloaded once a day by the Quarkus Registry client. Try removing it and see if there is a performance impact

**Comment by phillip-kruger** [comment]
Thanks @gastaldi .

@gsmet deleting this does make the initial `mvn quarkus:dev` slower as it's downloading this artifacts. Very rough estimate is it add +-2 seconds to the initial `quarkus:dev`. From then onward it's cached so no issue. 

Maybe you can test this on you laptop and see if it's noticeable for you. For me it was not.

**Comment by gastaldi** [comment]
@phillip-kruger can we somehow hit the registry when the UI to add an extension is opened?

**Comment by phillip-kruger** [comment]
@gastaldi - let me play around and l'll get back here.

**Review by anfnkst — APPROVED** [review]

**Comment by phillip-kruger** [comment]
@gastaldi @gsmet - ok I now moved that this happen when the add screen loads. So nothing on quarkus:dev anymore. I think we are good to go. 
loading

File: extensions/smallrye-health/deployment/src/test/java/io/quarkus/smallrye/health/test/PathMatchingHttpSecurityPolicyTest.java [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+    URL url;
+
+    private WebClient getClient() {
+        if (client == null) {
+            client = WebClient.create(vertx);
+        }
+        return client;
+    }
+
+    @ParameterizedTest
```
Comment (cescoffier) [inline extensions/smallrye-health/deployment/src/test/java/io/quarkus/smallrye/health/test/PathMatchingHttpSecurityPolicyTest.java:67]: Woot! I didn't know we could use parameterized tests when using the register extension approach. 

@mkouba, we discussed this recently.

File: extensions/vertx-http/deployment/src/test/java/io/quarkus/vertx/http/security/PathMatchingHttpSecurityPolicyTest.java [hunk 6]
```diff
@@ -190,25 +185,6 @@ public void testStaticResource(String path) {
         assurePathAuthenticated(path);
     }
 
-    @ParameterizedTest
```
Comment (cescoffier) [inline extensions/vertx-http/deployment/src/test/java/io/quarkus/vertx/http/security/PathMatchingHttpSecurityPolicyTest.java:193]: Why are they moved?

**Review by cescoffier — APPROVED** [review]
LGTM - I just wonder why the path tests have been moved around.

File: extensions/vertx-http/deployment/src/test/java/io/quarkus/vertx/http/security/PathMatchingHttpSecurityPolicyTest.java (hunk 6 above)
Comment (phillip-kruger) [reply extensions/vertx-http/deployment/src/test/java/io/quarkus/vertx/http/security/PathMatchingHttpSecurityPolicyTest.java:193]: So, here is the story on why they moved. The tests is in vertx-http, but (test only) need dependencies on openapi and health, both of which depend on vertx-http. The Windows build failed with this error, however the Linux build did not. I am not sure if the Linux build has some cache from a prev build (no-test) on CI, and Windows does not or why this only happens on Windows, but I could recreate it locally on a VM. It also seem that the failure in Windows is triggered by the new dependency added to the vertx-deploy module. I have no idea why. So to "fix" (sidestep) this, I moved the openapi test to openapi and the heath to health. This solved the issue, but i still can not say why this happened, only on Windows and triggered by the dependency.

Comment (phillip-kruger) [reply extensions/vertx-http/deployment/src/test/java/io/quarkus/vertx/http/security/PathMatchingHttpSecurityPolicyTest.java:193]: @cescoffier are you ok to merge with the explanation above ?

**Comment by cescoffier** [comment]
> LGTM - I just wonder why the path tests have been moved around.

@phillip-kruger gave me the answer. The tests (made by me a few years ago) were not located in their proper location - it was working _magically). It makes a lot more sense now.

**Comment by cescoffier** [comment]
Remove the backport label - this is a feature, not a bug fix.

**Pull request merged**